src/util/CancellationHandle.h

```diff
@@ -129,7 +129,10 @@ class CancellationHandle {
     using DurationType =
         std::remove_const_t<decltype(DESIRED_CANCELLATION_CHECK_INTERVAL)>;
 
-    if (state == CancellationState::CHECK_WINDOW_MISSED) {
+    bool windowMissed = state == CancellationState::CHECK_WINDOW_MISSED;
+    auto changed = cancellationState_.compare_exchange_strong(
+        state, CancellationState::NOT_CANCELLED, std::memory_order_relaxed);
+    if (windowMissed && changed) {
```

Comment:
I think `responsibleForReporting` or `notAlreadyHandled` is a semantically stronger alternative to `changed` as a name.

And the code
`if (! yourCurrentCondition) return ;`
makes everything less nested, in general I like such early bailouts if  they are simply and the nested code is rather long.